google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -62,7 +63,7 @@ public String toString() {
 
   final FirestoreImpl firestore;
 
-  private final List<WriteOperation> writes = new ArrayList<>();
+  private final List<WriteOperation> writes = new CopyOnWriteArrayList<>();
```

Comment:
I added comment indicating need for a thread safe collection.

I did not add a comment as to why CopyOnWriteArrayList was chosen. There are alternatives.

I followed advice from [go/java-practices/collection-implementations](https://goto.google.com/java-practices/collection-implementations).

Collections.synchronizedList(..) wrapper could also have been used.

Reply:
Note use of `Vector` is discouraged.

go/java-practices/choosing-collections#avoid

Reply:
Thanks for that link. Still, `CopyOnWriteArrayList` doesn't feel like the correct data structure here as it is optimized for the use case where reads dominate writes; however, the usage pattern here seems to be the exact opposite. This is mainly because appending an element to a `CopyOnWriteArrayList` is an O(n) operation (compared to ArrayList which is O(1) on average). So adding n elements to a `CopyOnWriteArrayList` is an O(n^2) operation. Can you confirm that this performance is acceptable?

Reply:
Also, changing it to a `CopyOnWriteArrayList` still leaves some race conditions in the code. Namely, there are several places that access the `writes` array multiple times seemingly under the assumption that it will not undergo any modifications between those accesses.

For example, this pattern is used often:

```java
writes.add(...);
return wrapResult(writes.size() - 1);
```

This is not thread-safe because the size of the `writes` array could change between the calls to `writes.add()` and `writes.size()` in the presence of concurrent modifications to `write`.

Also, in `commit()` it logs `writes.size()` then iterates over it in a separate step. This could be really confusing if there is an intervening modification to `writes` and the logged number of writes differs from the _actual_ number of writes that were performed.

Also, the `committed` member variable is another source of a data race.

Basically, to make this class truly thread safe will require more work than just swapping out ArrayList for a "thread-safe" list.

Reply:
Thank you for identifying concurrency problems.

Intent of original PR was to prevent corruption of `ArrayList` during proper usage.

Improper usage of calling write after commit wasn't really considered in scope. There is a check that already throws an exception telling developer not to do this. But we can strengthen this, to also throw an exception for concurrent case that can intermittently pass precondition.

I don't think the `writes.size() - 1` is a real problem because only the BulkWriter uses this and already has thread safety. But for the sake of completeness, I will add `synchronization` to UpdateBuilder as a second layer of safety and protection of future development.

Reply:
After discussing in person, changes were made that satisfy these concerns.